### ISSUE TITLE
batch create_entity error 400

### ISSUE THREAD

**Issue by kovapatrik** [SAP/python-pyodata#203]
Is it possible to create a batch of create_entity requests?
I've got an error 400 status code for an attempt, but I'm not sure if the library doesn't allow to do this, or our organization's SAP system.
I'm guessing the first one because I didn't find any example of doing this.

**Comment by phanak-sap** [comment]
Working with batch request is documented here:
https://github.com/SAP/python-pyodata/blob/master/docs/usage/advanced.rst

Should work, we use that. If you are having still issues, we would need much more information than HTTP 400 - something that could be ideally reproduced locally. At least enable logging to DEBUG level (same doc page) and provide the log file - scan first for credentials that you do not want to post to the internet.

**Comment by kovapatrik** [comment]
Yeah I saw that part of the documentation, I meant that there is no example for the create_entity function. I have something in my mind that could go wrong. I will check it tomorrow, and I will post if it's solved or not.

**Comment by kovapatrik** [comment]
I tried it and sadly it did not work. Could it be a setting on the server side which doesn't allow me to create a batch request? I tried it with get_entity methods too, and it didn't work.

**Comment by phanak-sap** [comment]
@kovapatrik Did you read [my comment - asking for some facts and details?](https://github.com/SAP/python-pyodata/issues/203#issuecomment-1027099708) I am sorry, but at the moment there is literally nothing we can do on our side. We cannot comment on configuration of unknown backend service (nor we can really support any backend service "to work") and there are no information on the error on the pyodata side - no stacktrace at least, no complete log, no test/sample code to reproduce the error somehow.

- Are you at least able to correctly create your entity with direct request, while failing inside batch request?
- Is your backend service even Odata V2 (and not for example v3/v4, which is not yet supported)?
- With what version of pyodata are you having the problem?
- Check existing better, more actionable issues here, with sample code provided. For example here - #187. This issue provided at least some starting point to investigate possible bug. Create therefore  some [reproducer code as shown here](https://github.com/phanak-sap/pyodata-issue-files/tree/master/%23187), that would for example initialize the pyodata with your metadata as local file and then provide the code to create your entity, so can be debugged the generated HTTP Request. 
But if it will look OK according to [odata specification ](https://www.odata.org/documentation/odata-version-2-0/batch-processing/) then it is just a guessing game why the server returns HTTP 400, that's for investigation on your end. 

Understand it from my point of view. If I wanted to be a little sarcastic, I could reply instead with just "I tried to fix your problem and it didn't work too" :)

**Comment by kovapatrik** [comment]
I didn't even expect you to solve my problem with the information I was given so you don't have to be sarcastic. I just asked a question about a backend service I have no experience with, before we thinking about a bug in the pyodata library itself. I thought an SAP service is standard so there are standard possible configurations. But in any case, here is the code I'm using: 

`
s = requests.Session()
s.auth = HTTPKerberosAuth(mutual_authentication=OPTIONAL)
s.headers = {'x-csrf-token': 'fetch'}
s.verify = ca_bundle_location

res = s.get(SERVICE_URL)
csrf_token = res.headers.get("x-csrf-token", "")
s.headers.update({"x-csrf-token": csrf_token})

client = pyodata.Client(SERVICE_URL, s)

batch = client.create_batch()

rq1 = client.entity_sets.LeaveRequestSet.create_entity().set(**d1)
rq2 = client.entity_sets.LeaveRequestSet.create_entity().set(**d2)

batch.add_request(rq1)
batch.add_request(rq2)

batch.execute()
`

For the batch.execute() command I get this error:
`HttpError: HTTP POST for multipart request 1989_1874_3537 failed with status code 400`

And [here](https://gist.github.com/kovapatrik/249841aa92abccc213ea562a50da84e8) is the debug output.

**Comment by phanak-sap** [comment]
Hi @kovapatrik, thx for provided info.

The HTTP 400  Bad requests is obvious. The correct creation without batch requests is a valuable fact.

What catched my eye in the provided python script is missing changeset wrapping. My current understanding of odata is that each create call should be in its own changeset = in your example 1 batch request, inside 2 changesets. Your script should IMHO work correctly if it would be just two entity queries, then the changesets should not be needed.

Please try to modify it according to second sample from [pyodata documentation](https://pyodata.readthedocs.io/en/latest/usage/advanced.html?highlight=batch#batch-requests)

I hope it will help and it is just that simple. If not, we will need to investigate deeper and consider this a possible bug.

links:
- https://pyodata.readthedocs.io/en/latest/usage/advanced.html?highlight=batch#batch-requests
- https://blogs.sap.com/2018/05/06/batch-request-in-sap-gateway/
- https://www.odata.org/documentation/odata-version-2-0/batch-processing/

**Comment by phanak-sap** [comment]
Note: I also noticed that we still have pending Enhancement for handling partial success scenario - issue #64

**Comment by kovapatrik** [comment]
The changset wrapping indeed solves the problem. But I had to wrap each create_entity request within it's own changeset, because if I added 2 requests into 1 changeset, the servers responded with an error like the server only allows 1 change per changeset.

_I didn't meet with odata before, so IMO it would be nice if there is an example in the docs also for create_entity requests in batch including the changeset wrapping._

So thanks for the help!

My last question: performance-wise, is it better to send requests in batch or it doesn't matter?

**Comment by kovapatrik** [comment]
Oh I see you answered my question before I even asked :D I will look up the module you linked, thank you!

**Comment by phanak-sap** [comment]
1. if the examples in Advanced.rst were not helpful for you from start, I will be happy to review PR with update of this file with example which would be for you more understandable.
2. _"Example of batch request that contains simple entity query as well as changest consisting of two requests for entity modification"_  - yeah that a bit expects you know about Batches more than just what is stated in this line only. It is true that user of pyodata package, even that is somehow shielded from for example the precise URL generation, is still somehow expected to understand basics of odata protocol. 
3. It could be a nice to have enhancement tho, to check the Batch content before sent, if changesets are missing and needed.

**Comment by phanak-sap** [comment]
> My last question: performance-wise, is it better to send requests in batch or it doesn't matter?

There will be definitely some difference between X create requests sent in a simple loop (due synchronous wait on response) and sending the same X create requests in a Batch. But it primarily saves networking resources (for example in browser apps, Fiori), from time standpoint, the server will still need to use basically same resources to actually create the entity somewhere in DB. 

The gevent/multihreading will remove the waiting time a bit, so for big test data creation script it will end definitely sooner, that's for sure. But check first if it is even worth the hassle :)  Maybe in a simple loop you will still create enough data in seconds or minutes and no need to actually optimize.  Or you want to create huge batch of data, but no problem if executed overnight, once a year. I

**Comment by phanak-sap** [comment]
> The changeset wrapping indeed solves the problem. But I had to wrap each create_entity request within it's own changeset, because if I added 2 requests into 1 changeset, the server responded with an error like the server only allows 1 change per changeset.

Yes, this was exactly what I meant, that this is IMHO mandatory in the odata protocol. One batch, two changesets in your example.

**Comment by kovapatrik** [comment]
It will be likely used once per month so I think I won't create any threading and use the simple create requests command.

For the docs change: Now I understand that I have to know something about the odata protocol before using pyodata, so I guess if I knew a create request must be in a changeset, then the docs are clear on how to use changesets.

**Comment by phanak-sap** [comment]
Closing as resolved.
